Kuadrant/limitador-operator
# Pass through Pod Affinity Rules

Allow Pod Affinity Rules to be defined in the Limitador CR which will be past to the Limitador deployment CR and place in `spec.template.spec.affinity`.

This change address #76 

# Validation 
* Run `make local-setup`
* Create the Limitador CR with out Pod Affinity Rules
```sh
echo "
   apiVersion: limitador.kuadrant.io/v1alpha1
   kind: Limitador
   metadata:
     name: limitador-sample
   spec:
     replicas: 3
     limits:
       - conditions: ["get_toy == 'yes'"]
         max_value: 2
         namespace: toystore-app
         seconds: 30
         variables: []
   " | kubectl apply -f -
```
* **Expected:** deployment has no Affinity Rules
```sh
kubectl get deployment limitador-sample -o jsonpath='{.spec.template.spec.affinity}' -n default | jq
```
* update the Limitador CR to contain Pod Affinity Rules.
```sh
echo "
apiVersion: limitador.kuadrant.io/v1alpha1
kind: Limitador
metadata:
  name: limitador-sample
spec:
  replicas: 3
  affinity:
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
        - weight: 100
          podAffinityTerm:
            labelSelector:
              matchLabels:
                app.kubernetes.io/name: limitador
            topologyKey: kubernetes.io/hostname
        - weight: 99
          podAffinityTerm:
            labelSelector:
              matchLabels:
                app.kubernetes.io/name: limitador
            topologyKey: topology.kubernetes.io/zone
  limits:
    - conditions: ["get_toy == 'yes'"]
      max_value: 2
      namespace: toystore-app
      seconds: 30
      variables: []
   " | kubectl apply -f -
```
* **Expected:** deployment contains Affinity Rules
```sh
kubectl get deployment limitador-sample -o jsonpath='{.spec.template.spec.affinity}' -n default | jq
```
* Update the Affinity Rules
```sh
echo "
apiVersion: limitador.kuadrant.io/v1alpha1
kind: Limitador
metadata:
  name: limitador-sample
spec:
  replicas: 3
  affinity:
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
        - weight: 50
          podAffinityTerm:
            labelSelector:
              matchLabels:
                app.kubernetes.io/name: limitador
            topologyKey: kubernetes.io/hostname
        - weight: 50
          podAffinityTerm:
            labelSelector:
              matchLabels:
                app.kubernetes.io/name: limitador
            topologyKey: topology.kubernetes.io/zone
  limits:
    - conditions: ["get_toy == 'yes'"]
      max_value: 2
      namespace: toystore-app
      seconds: 30
      variables: []
   " | kubectl apply -f -
```
* **Expected:** deployment Affinity weights to be modified.
```sh
kubectl get deployment limitador-sample -o jsonpath='{.spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[*].weight}' -n default | jq
```
* Removing Affinity Rules from Limitador CR removes rules from deployment CR
```sh
echo "
   apiVersion: limitador.kuadrant.io/v1alpha1
   kind: Limitador
   metadata:
     name: limitador-sample
   spec:
     replicas: 3
     limits:
       - conditions: ["get_toy == 'yes'"]
         max_value: 2
         namespace: toystore-app
         seconds: 30
         variables: []
   " | kubectl apply -f -
```
* **Expected:** deployment has no Affinity Rules
```sh
kubectl get deployment limitador-sample -o jsonpath='{.spec.template.spec.affinity}' -n default | jq
```
